File: Programming_I/3rd_cut/Workshop/Ts/quicksort.ts
import * as beauty from "./Controllers";

function main(): number {
  let array = [4, 2, 6, 0, 12, 42, 8];
  let arraySize = array.length;

  console.clear();
  beauty.centerPrint("Unsorted array", 5);
  beauty.arrayTable(array, 6);

  beauty.centerPrint("Quicksorted array", 10);
  quicksort(array, 0, arraySize - 1);
  beauty.arrayTable(array, 11);

  return 0;
}

function arrayPartition(
  array: Array<number>,
  init: number,
  limit: number
): number {
  let j: number,
    swap: number,
    i = init - 1;
  let pivot = array[limit];

  for (j = init; j < limit; j++) {
    if (array[j] <= pivot) {
      i++;
      swap = array[i];
      array[i] = array[j];
      array[j] = swap;
    }
  }

  swap = array[i];
  array[i] = array[j];
  array[j] = swap;
  return i + 1;
}

function quicksort(array: Array<number>, init: number, limit: number): void {
  let partition_index: number;

  if (init < limit) {
    partition_index = arrayPartition(array, init, limit);

    quicksort(array, init, partition_index - 1);
    quicksort(array, partition_index + 1, limit);
  }
}

main();

File: Programming_I/3rd_cut/Workshop/Cpp/quicksort.cpp
#include "controllers.hh"

int Partition(int array[], int init, int limit);
int QuickSort(int array[], int init, int limit);

int main()
{
    int array[] = {4, 2, 6, 0, 12, 42, 8};
    int array_size = sizeof(array) / sizeof(*array);

    SetWindow('n');
    system("cls");

    CenterPrint("Unsorted array", 5);
    ArrayTable(array, array_size, 6);

    CenterPrint("Quicksorted array", 10);
    QuickSort(array, 0, array_size - 1);
    ArrayTable(array, array_size, 11);
    std::cout << "\n";
    return 0;
}

int Partition(int array[], int init, int limit)
{
    int j, i = init - 1;
    int pivot = array[limit];
    int changer;

    for (j = init; j < limit; j++)
        if (array[j] <= pivot)
        {
            i++;
            changer = array[i];
            array[i] = array[j];
            array[j] = changer;
        }
    changer = array[i+1];
    array[i+1] = array[limit];
    array[limit] = changer;
    return i + 1;
}

int QuickSort(int array[], int init, int limit)
{
    int partition_index;

    if (init < limit)
    {
        partition_index = Partition(array, init, limit);

        QuickSort(array, init, partition_index-1);
        QuickSort(array, partition_index+1, limit);
    }
}